pygfx/objects/_base.py

```diff
@@ -294,63 +257,65 @@ def receive_shadow(self, value):
         self._store.receive_shadow = bool(value)
 
     @property
-    def parent(self):
-        """Object's parent in the scene graph (read-only).
-        An object can have at most one parent.
-        """
-        return self._parent_ref and self._parent_ref()
+    def material(self):
+        """Wheter is object is rendered or not. Default True."""
+        return self._store.material
+
+    @material.setter
+    def material(self, material):
+        self._store.material = material
 
     @property
-    def children(self):
-        """The child objects of this wold object (read-only tuple).
-        Use ``.add()`` and ``.remove()`` to change this list.
-        """
-        return tuple(self._children)
+    def parent(self) -> "WorldObject":
+        if self._parent is None:
+            return None
+        else:
+            return self._parent()
+
+    @parent.setter
+    def parent(self, value: "WorldObject"):
+        if value is None:
+            self._parent = None
+        else:
+            self._parent = weakref.ref(value)
 
     def add(self, *objects, before=None):
         """Adds object as child of this object. Any number of
         objects may be added. Any current parent on an object passed
         in here will be removed, since an object can have at most one
         parent.
-        If ``before`` argument is given (and present in children), then
-        the items are inserted before the given element.
+        If ``before`` argument is given, then the items are inserted before the
+        given element.
+
         """
-        idx = len(self._children)
-        if before:
-            try:
-                idx = self._children.index(before)
-            except ValueError:
-                pass
+
+        obj: WorldObject
         for obj in objects:
-            assert isinstance(obj, WorldObject)
-            # orphan if needed
-            if obj._parent_ref is not None:
-                obj._parent_ref().remove(obj)
-            # attach to scene graph
-            obj._parent_ref = weakref.ref(self)
-            self._children.insert(idx, obj)
-            idx += 1
-            # flag world matrix as dirty
-            obj._matrix_world_dirty = True
-        return self
+            # added for backwards compatibility
+            # (I feel like we should raise instead if the object is already part
+            # of the scene graph)
+            if obj.parent is not None:
+                obj.parent.remove(obj)
+
+            if before is not None:
+                idx = self.children.index(before)
+            else:
+                idx = len(self.children)
```

Comment:
Yes, because the position of `before` may change if an object being added did already exist on the parent and that object was located before `before`.

On that note, why does the order of children matter? I would have thought that a child is a child no matter which siblings come before or after it.

Reply:
* https://github.com/fork-tongue/collagraph depends on this feature
* If nothing else, the order of children has an influence on rendering order

Reply:
> the position of before may change if an object being added did already exist on the parent and that object was located before before.

That's a good point. But this gets complicated quickly, and it would be good to have some tests to make sure the behavior is correct. Shall we address this in a separate PR?

Reply:
I think it makes sense to pick it up here. I will have to add new unit tests for the transform classes anyway and potentially add some coverage for the world object if I notice things that break and that aren't covered while refactoring on pylinalg. I think this would fall into that category.